source/common/event/scaled_range_timer_manager.cc

```diff
@@ -0,0 +1,231 @@
+
+  void onMinTimerComplete() {
+    ENVOY_LOG_MISC(info, "min timer complete for {}", static_cast<void*>(this));
+    ASSERT(absl::holds_alternative<WaitingForMin>(state_));
+    WaitingForMin& waiting = absl::get<WaitingForMin>(state_);
+
+    if (waiting.scalable_duration_ < std::chrono::milliseconds::zero()) {
+      trigger();
+    } else {
+      state_.emplace<ScalingMax>(manager_.activateTimer(waiting.scalable_duration_, *this));
+    }
+  }
+
+  ScaledRangeTimerManager& manager_;
+  const TimerCb callback_;
+  const TimerPtr min_duration_timer_;
+
+  absl::variant<Inactive, WaitingForMin, ScalingMax> state_;
+  const ScopeTrackedObject* scope_;
+};
+
+ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher)
+    : dispatcher_(dispatcher), scale_factor_(1.0) {}
+
+ScaledRangeTimerManager::~ScaledRangeTimerManager() {
+  // Scaled timers created by the manager shouldn't outlive it. This is
+  // necessary but not sufficient to guarantee that.
+  ASSERT(queues_.empty());
+}
+
+RangeTimerPtr ScaledRangeTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<RangeTimerImpl>(callback, *this);
+}
+
+void ScaledRangeTimerManager::setScaleFactor(double scale_factor) {
+  const MonotonicTime now = dispatcher_.approximateMonotonicTime();
+  scale_factor_ = DurationScaleFactor(scale_factor);
+  for (auto& queue : queues_) {
+    resetQueueTimer(*queue, now);
+  }
+}
+
+ScaledRangeTimerManager::Queue::Item::Item(RangeTimerImpl& timer, MonotonicTime active_time)
+    : timer_(timer), active_time_(active_time) {}
+
+ScaledRangeTimerManager::Queue::Queue(std::chrono::milliseconds duration,
+                                      ScaledRangeTimerManager& manager, Dispatcher& dispatcher)
+    : duration_(duration),
+      timer_(dispatcher.createTimer([this, &manager] { manager.onQueueTimerFired(*this); })) {}
+
+ScaledRangeTimerManager::ScalingTimerHandle::ScalingTimerHandle(Queue& queue,
+                                                                Queue::Iterator iterator)
+    : queue_(queue), iterator_(iterator) {}
+
+ScaledRangeTimerManager::DurationScaleFactor::DurationScaleFactor(double value)
+    : value_(std::max(0.0, std::min(value, 1.0))) {}
+
+MonotonicTime ScaledRangeTimerManager::computeTriggerTime(const Queue::Item& item,
+                                                          std::chrono::milliseconds duration,
+                                                          DurationScaleFactor scale_factor) {
+  return item.active_time_ +
+         std::chrono::duration_cast<MonotonicTime::duration>(duration * scale_factor.value());
+}
+
+ScaledRangeTimerManager::ScalingTimerHandle
+ScaledRangeTimerManager::activateTimer(std::chrono::milliseconds duration,
+                                       RangeTimerImpl& range_timer) {
+  ASSERT(dispatcher_.isThreadSafe());
+  auto it = queues_.find(duration);
+  if (it == queues_.end()) {
+    auto queue = std::make_unique<Queue>(duration, *this, dispatcher_);
```

Comment:
Let me try to explain this code back to you to see if I understand it:

1) Every range timer has an individual min timer that fires independently.
2) Once the min timer fires, if there is a gap between min and max, we take the duration of the min/max delta, and put it in a queue unique to that delta. The assumption here is that there will be a small number of min/max configured so the number of queues is not that big.
3) Each queue has a timer that fires for the min/max duration scaled, and flushes all timers that have expired, and then resets for the next duration that needs to fire.

If so that makes sense, but per my other comments can you add a bunch more comments in these two files?

Reply:
Yep, that's correct. Added plenty of comments.